source/pdo_sqlsrv/pdo_stmt.cpp

```diff
@@ -1067,6 +1067,30 @@ int pdo_sqlsrv_stmt_next_rowset( _Inout_ pdo_stmt_t *stmt TSRMLS_DC )
 
         SQLSRV_ASSERT( driver_stmt != NULL, "pdo_sqlsrv_stmt_next_rowset: driver_data object was null" );
 
+        // Return the correct error in case the user calls nextRowset() on a null result set. 
+        // Null means that SQLNumResultCols() returns 0 and SQLRowCount is not return > 0. But first 
+        // check that the statement has been executed and that we are not past the end of a non-null 
+        // result set to make sure the user gets the correct error message. These checks are also 
+        // done in core_sqlsrv_next_result(), but we cannot check for null results there because that
+        // function can be called without calling this one, and SQLSRV_ERROR_NO_FIELDS can then
+        // be triggered incorrectly. 
+        CHECK_CUSTOM_ERROR( !driver_stmt->executed, driver_stmt, SQLSRV_ERROR_STATEMENT_NOT_EXECUTED ) {
+            throw core::CoreException();
+        }
+
+        CHECK_CUSTOM_ERROR( driver_stmt->past_next_result_end, driver_stmt, SQLSRV_ERROR_NEXT_RESULT_PAST_END ) {
+            throw core::CoreException();
+        }
+        
+        // Now make sure the result set is not null. 
+        bool has_result = core_sqlsrv_has_any_result( driver_stmt );
+
+        if(!driver_stmt->fetch_called){
+            CHECK_CUSTOM_ERROR( !has_result, driver_stmt, SQLSRV_ERROR_NO_FIELDS ) {
+                throw core::CoreException();
+            }
+        }
```

Comment:
I don't quite understand. I thought PSOStatement::nextRowset should only be called on a fetched statement. The condition for if says fetched is NOT called, and throws an exception if there is no result. How can there any result if the statement was never fetched?

Reply:
nextRowset() *should* only be called on a fetched statement, but there is nothing stopping the user from calling nextRowset immediately after executing, and we need to handle that case. In issue 507 it occurs when going into the 'else' part of stored procedure that returns a result set in the 'if' but not the 'else' branch, so it can happen that user does call nextRowset first accidentally via a stored procedure. 

This code catches the case where nextRowset() is called before fetch on a null result set, giving the appropriate error message; before, there would be no error given. In the case where nextRowset is called before fetch on non-null result set, core_sqlsrv_has_any_result will return true and the error is not triggered. In the event that nextRowset is called after fetch() and the result set is null, that will be handled in core_sqlsrv_next_result() (or by the check for past_next_result_end above). I think that covers all the relevant cases...

Reply:
yea I was confused what happens when fetch isn't called and has_result is true, but now it's more clear. Can you put a comment saying if fetch_called is false and has_result is true, core_sqlsrv_next_result will handle it.